index.js

```diff
@@ -30,26 +43,23 @@ class BindingComponent extends React.Component {
         val = this.props.bindStateType(val);
       }
     }
-    const names = this.props.bindStateName.split('.');
-    if (names.length === 1) {
-      this.props.bindStateCtx.setState({
-        [names[0]]: val
-      });
-    } else {
-      const newState = this.props.bindStateCtx.state[names[0]];
-      const subnames = names.slice(1);
-      subnames.reduce((curr, name, index) => {
-        if (index === subnames.length - 1) {
-          curr[name] = val;
-        } else {
-          curr = curr[name];
-        }
-        return curr;
-      }, newState);
-      this.props.bindStateCtx.setState({
-        [names[0]]: newState
-      });
-    }
+
+    this.props.bindStateCtx.setState((prevState) => {
+      let pathReversed;
+      let overlay = val;
+      let pathIndex = 0;
+      try {
+        pathReversed = this.props.bindStateName.split('.').reverse();
+      } catch (error) {
+        return prevState;
+      }
+      do {
+        let subpath = pathReversed[pathIndex];
+        overlay = { [subpath]: overlay };
+        pathIndex++;
+      } while (pathIndex < pathReversed.length);
+      return assign(prevState, overlay);
```

Comment:
Here we needn't care about if *setState* does merge deeply, because this operation does only do a variable assignment and trigger UI updates. So the first function in *setState* is required to return the fields to be re-assigned and react triggers UI updates by itself.

If the above description is not clear, let me show a little bit code for you:

```js
setState((prevState, currProps) => {
  let foo = prevState.foo;
  foo.bar.x.zoo = 10;
  return { foo };
}, () => {
  // here we will get the updated correct state object.
});
```